### ISSUE TITLE
[dynamo] Slow compile times for optimizers due to for loops

### ISSUE THREAD

**Issue by jon-chuang** [pytorch/pytorch#110506]
### 🐛 Describe the bug

Dynamo tracing time is:
- 70 seconds on 200 param Adam
- 162 seconds on 1000 param SGD

As identified in https://github.com/pytorch/pytorch/pull/110353#issuecomment-1746729070, this is due to dynamo needing to trace an expensive for loop.

If instead this for loop can be written in a way that can be easily traced (e.g. by tracing a `map` over a single lambda/similar to foreach over the optimizer main loop) then we are likely to speedup compilation times across all optimizers by a significant factor.

Example here: https://github.com/pytorch/pytorch/blob/31d635803b8d72433ea275d3c36bf829b158d5ec/torch/optim/sgd.py#L39
Example dynamo logs (which trace the same computation ad infinitum):
```
[2023-10-04 08:14:53,891] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST grad [ListIteratorVariable()]
[2023-10-04 08:14:53,891] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR is_sparse [ListIteratorVariable(), TensorVariable()]
[2023-10-04 08:14:53,892] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE YIELD_VALUE None [ListIteratorVariable(), ConstantVariable(bool)]
[2023-10-04 08:14:53,892] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE POP_TOP None [ListIteratorVariable(), ConstantVariable(NoneType)]
[2023-10-04 08:14:53,892] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE JUMP_ABSOLUTE 4 [ListIteratorVariable()]
[2023-10-04 08:14:53,892] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE FOR_ITER 18 [ListIteratorVariable()]
[2023-10-04 08:14:53,895] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE STORE_FAST grad [ListIteratorVariable(), TensorVariable()]
[2023-10-04 08:14:53,895] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST grad [ListIteratorVariable()]
[2023-10-04 08:14:53,895] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR is_sparse [ListIteratorVariable(), TensorVariable()]
[2023-10-04 08:14:53,896] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE YIELD_VALUE None [ListIteratorVariable(), ConstantVariable(bool)]
[2023-10-04 08:14:53,896] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE POP_TOP None [ListIteratorVariable(), ConstantVariable(NoneType)]
[2023-10-04 08:14:53,896] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE JUMP_ABSOLUTE 4 [ListIteratorVariable()]
[2023-10-04 08:14:53,896] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE FOR_ITER 18 [ListIteratorVariable()]
[2023-10-04 08:14:53,900] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE STORE_FAST grad [ListIteratorVariable(), TensorVariable()]
[2023-10-04 08:14:53,900] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST grad [ListIteratorVariable()]
[2023-10-04 08:14:53,900] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR is_sparse [ListIteratorVariable(), TensorVariable()]
[2023-10-04 08:14:53,901] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE YIELD_VALUE None [ListIteratorVariable(), ConstantVariable(bool)]
[2023-10-04 08:14:53,901] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE POP_TOP None [ListIteratorVariable(), ConstantVariable(NoneType)]
[2023-10-04 08:14:53,901] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE JUMP_ABSOLUTE 4 [ListIteratorVariable()]
[2023-10-04 08:14:53,901] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE FOR_ITER 18 [ListIteratorVariable()]
```

CC: @mlazos @jansel 

### Repro
```python
import time
import torch
from torch.optim import Adam, SGD

def compile_opt(opt_compiled):
    torch._dynamo.eval_frame.TorchPatcher.patch()

    step_fn = opt_compiled.step.__wrapped__
    def fn():
        step_fn(opt_compiled)

    return torch.compile(fn, backend="inductor", fullgraph=True)

optim_cls = SGD
NUM_PARAMS = 1000
kwargs = { "lr": 0.01, "foreach": True }

torch._dynamo.reset()
# torch._inductor.metrics.reset()
input = torch.ones([10, 10], device="cuda:0")
model = torch.nn.Sequential(
    *[torch.nn.Linear(10, 10, device="cuda:0") for _ in range(NUM_PARAMS)]
)

input = torch.ones([10, 10], device="cuda:0")
model(input).sum().backward()
opt_compiled = optim_cls(model.parameters(), **kwargs)
compiled_step = compile_opt(opt_compiled)

with torch.set_grad_enabled(False):
    start_time = time.time()
    compiled_step()
    print("compile opt took: %s seconds", time.time() - start_time)
```

### Versions

main

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @ezyang @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @Xia-Weiwen @aakhundov

**Comment by jon-chuang** [comment]
@mlazos you're right. That was a misdiagnosis.

However, there are many other for loops that are hurting tracing. Examples:

https://github.com/pytorch/pytorch/blob/21d77bcf808d076f81b5e885a8ce6ca20a08dbed/torch/optim/sgd.py#L283
changing this line to use `has_sparse_grad` instead: 12s -> 9s (NUM_PARAMS=300), 162s-> 78s (NUM_PARAMS=1000).

This seems to be due to `any`?

Will continue to investigate Adam (which as noted, is actually far worse).

**Comment by jon-chuang** [comment]
@mlazos, based on the above results, I'm surprised any of `Adam`, `Adagrad` and any other optimizer that uses list comprehensions to coerce real are enabled at all. Any optimizer that uses this takes 100+ seconds on 200 params.

I'll try to investigate moving these into the `_init_group` as a first priority.

**Comment by mlazos** [comment]
In the end, my strategy with these checks is only check them if is_compiling is False, and then add the check into the optimizer variable. I think this might be the best approach vs moving things into _init_group

**Comment by jon-chuang** [comment]
> I think this might be the best approach vs moving things into _init_group

Is there any disadvantage to moving things into `_init_group`?

**Comment by Chillee** [comment]
![A SVG file from running with TORCH_COMPILE_CPROFILE=1](https://github.com/pytorch/pytorch/assets/6355099/8c9e6d00-5018-4c9b-b046-c5e7b9329493)

Seems like a lot of it is bottlenecked by Inductor. Fake tensor caching will eliminate about 20%, but I feel like we're also just processing too many nodes in Inductor for no reason?

**Comment by jon-chuang** [comment]
Yep, as of main today a most of the dynamo bottleneck is no longer present, due to https://github.com/pytorch/pytorch/pull/110613.  As for the current bottleneck in inductor, it seems a lot of time is spent in lowering. There are a lot of nodes as it includes all of the model parameters and state - so usually a constant multiple more params than the model itself.

But the actual compute graphs have a lot of repetition. A lot of overhead has been alleviated with use of foreach APIs but not completely.

Another issue is the debuggability is nearly zero, as the var names which are derived by concatting original buffer names are 1000s of characters long - which also seems worth improving.

**Comment by mlazos** [comment]
@jon-chuang we generally don't care about non-foreach anymore, it just takes too long to trace, and even fixing that, it is tricky for inductor to fuse the large graph properly into a single kernel without extra structure, which is why I switched to foreach. 

You're correct about the super linear behavior in inductor, from my investigation it looks like this is due to FX tracing the loop body of copy_ ops that are introduced during functionalization. In our case there can be 1000s of these for each mutating foreach op (one for each output tensor getting mutated). The issue is that after tracing these loop bodies a lot of these copies get fused into upstream foreach ops in the scheduler anyway.  My plan is in the pattern matcher to replace the single copy_ ops with a foreach_copy_ which will allow the unfused copy_ ops during lowering to be combined with upstream foreach ops, this will reduce the number of single copy_ buffers sent to the scheduler, and there will be a lot fewer loop bodies to trace.

An alternative solution (or possibly additive solution) is to have a fast-path for tracing the FX loop bodies when the index expressions for dims and strides are simple

**Comment by janeyx99** [comment]
@mlazos @jon-chuang While compilation time may not be as high priority for non-foreach implementations (compared to foreach), we still do care about supporting them as much as we can 😅, and if you have any ideas on low hanging fruit, we would be happy to consider. But yes, we are working on unblocking the foreach use case first as that is the default for most optimizers.

**Comment by jon-chuang** [comment]
While we have mitigated the effects for the typical path (no has_complex), when there are complex, we can benefit from converting the tensors themselves to real within dynamo, and then handing the references to the views as inputs to dynamo. I'm not sure if this breaks things though (I think views only share the underlying storage, but do they share other mutations e.g. metadata mutations?)

**Comment by mlazos** [comment]
> A SVG file from running with TORCH_COMPILE_CPROFILE=1
> 
> ![A SVG file from running with TORCH_COMPILE_CPROFILE=1](https://user-images.githubusercontent.com/6355099/273351680-8c9e6d00-5018-4c9b-b046-c5e7b9329493.svg)
> 
> Seems like a lot of it is bottlenecked by Inductor. Fake tensor caching will eliminate about 20%, but I feel like we're also just processing too many nodes in Inductor for no reason?

A quick update on this: 
https://github.com/pytorch/pytorch/pull/112440 shows a prototype of how to reduce the number of buffers getting sent to inductor by about half, this helped with 1k params with Adam by about 35%. This + fake tensor caching looks like it will get 1k params to between 110~120 seconds. 

I think this approach is really promising, and maybe finding other ways to shortcut reindexing passes for contiguous tensors could help even more. In the end, getting as much done as possible during lowering looks to be the best way to help the scheduler. I also have some ideas around optimizing computing dependencies after fusing foreach which might help too but I haven't profiled inductor alone yet. If anyone wants to investigate some of those ideas lmk.

**Comment by anijain2305** [comment]
Adding indcutor tag because the compile time opportunities now are in Inductor.

**Comment by mlazos** [comment]
I'm going to close this, since dynamo tracing time is 6s now on 200 param Adam. If the inductor compile times are issues for users feel free to reopen - but this has been sped up quite a lot in 23H2, stands at ~45s for 200 params. I noticed in this repro, since Linears were used as the example we were actually running with twice the number of parameters indicated with `NUM_PARAMS` because of the bias.
